/*
  This file holds scan message types, for sending the results of individual
  scans between afspm nodes.
*/

syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "geometry.proto";

/* Enum indicating the current state of scanning */
enum ScanState {
  SS_UNDEFINED = 0;
  SS_MOVING = 1;  // Moving, not scanning
  SS_SCANNING = 2;
  SS_FREE = 3; // Neither scanning nor moving, free to do something else
}


/* Holds data, shape, and units. */
message ScanData{
  optional Size2u shape = 1;
  optional string units = 2;
  repeated double values = 3; // Expect [y][x]
}

/*
  ScanParameters contains the physical coordinates of the scan, as well as
  the shape of the 2D data array captured.
*/
message ScanParameters2d {
  optional Rect2d spatial_roi = 1;
  optional google.protobuf.Timestamp timestamp = 2;
  optional string name = 3; // Name to attribute to this data array
}

/* Scan contains the parameters and data array from an individual scan. */
message Scan2d {
  optional ScanParameters2d params = 1;
  optional ScanData data = 2;
}
